File: files/CH-100.py
import pandas as pd
import numpy as np

path = "CH-100 Manage Duplicate Values.xlsx"
input = pd.read_excel(path, usecols="B", skiprows=1, dtype=str)
test  = pd.read_excel(path, usecols="D", skiprows=1, dtype=str)
test.columns = test.columns.str.replace('.1', '')

result = input.copy()
result['row'] = result.groupby('Product ID').cumcount()
result['nrows'] = result.groupby('Product ID')['Product ID'].transform('size')
result['letter'] = np.where(result['nrows'] > 1, result['row'].apply(lambda x: chr(x + 65)), '')
result['Product ID'] = result['Product ID'] + result['letter']
result = result[['Product ID']]

print(result.equals(test))  # True

File: files/100-199/100/CH-100.R
library(tidyverse)
library(readxl)

path = "files/CH-100 Manage Duplicate Values.xlsx"
input = read_excel(path, range = "B2:B15")
test  = read_excel(path, range = "D2:D15")

result = input %>%
  mutate(row = row_number(), 
         n = n(), 
         .by = `Product ID`) %>%
  mutate(letter = ifelse(n > 1, LETTERS[row], "")) %>%
  select(`Product ID`, letter) %>%
  unite("Product ID", `Product ID`, letter, sep = "")

identical(result, test)
#> [1] TRUE